docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,234 @@
+of a package. That is, it defines the original (upstream) package, the new
+(downstream) package, and the changes (mutations) that need to be made to
+transform the upstream into the downstream. It also allows the user to specify
+policies around adoption, deletion, and update of package revisions that are
+under the control of the package variant controller.
+
+![Basic Package Variant](PackageVariant%20-%201.png)
+
+Note that *proposal* and *approval* are not handled by the package variant
+controller. Those are left to humans or other controllers. The exception is the
+proposal of deletion (there is no concept of a "Draft" deletion), which the
+package variant control will do, depending upon the specified deletion policy.
+
+### Package Revision Metadata
+
+The package variant controller utilizes Porch APIs. This means that it is not
+just doing a `clone` operation, but in fact creating a Porch `PackageRevision`
+resource. In particular, that resource can contain Kubernetes metadata that is
+not part of the package as stored in the repository.
+
+Some of that metadata is necessary for the management of the PackageRevision by
+the PackageVariant controller - for example, the `ownerRef` indicating which
+`PackageVariant` created the `PackageRevision`. These are not in control of the
+user. However, the `PackageVariant` resource does make the annotations and
+labels of the `PackageRevision` available as values that may be controlled
+during the creation of the `PackageRevision`. This can assist in additional
+automation workflows.
+
+## Introducing Variance
+Just cloning is not that interesting, so the `PackageVariant` resource also
+allows you to control various ways of mutating the original package to create
+the variant.
+
+### Package Context
+[Note: I believe right now we *always* produce `--for-deployment` package
+revisions out of package variant controller. We will likely need to allows some
+control over that, as we already have a use case in Nephio to use a
+PackageVariant to create a package intended as a base for fan out]
+
+Every kpt package that is designated `--for-deployment` will contain a
+ConfigMap called `kptfile.kpt.dev`. Kpt (or Porch) will automatically add a
+key `name` to the ConfigMap data, with the value of the package name. This
+ConfigMap can then be used as input to functions in the Kpt function pipeline.
+
+This process holds true for package revisions created via the package variant
+controller as well. Additionally, the author of the `PackageVariant` resource
+can specify additional key-value pairs to insert into the package context [note:
+this additional functionality is proposed here but not yet implemented].
+
+### KRM Function Calls
+[Note: not yet implemented]
+TODO(johnbelamaric): describe adding a KRM function pipeline to the
+PackageVariant to allow arbitrary mutations
+- question: should these allow adding to the package pipeline a la `--save`? Is
+  that a *separate* pipeline?
```

Comment:
I like the idea of adding a KRM function pipeline to be added to the PackageVariant, and having that saved to the Kptfile. But there are two questions for me:
1. Does adding it to the package pipeline need to be an option, or should we always add it (or at least add it by default)? Would it be common to run these functions as a one-off thing and not save it to the Kptfile? If they add more resources to the Kptfile later, I would imagine that they'd want to rerun all the same functions again, which is what the Kptfile pipeline helps with.
2. Could you clarify what you mean by a "separate" pipeline? Not sure I fully understand the advantages of having it separate from the package pipeline

Reply:
1. I am not sure we know yet. In some ways `PackageVariant` is trying to emulate what a user might do, which could include imperative calls to KRM functions. That's where this question comes from.
2. We have talked in the past that there are cases (like managing app config), where sets of functions may need to be run at different times or for different operations. So this question is in that context. I am not sure of the answer. 

For both of these, we may want to just pick the simplest option and see if it ends up being sufficient or not.

Reply:
One possible use would be for functions that require network or volume access. For example, we may want to add the ability for `PackageVariant` to render a Helm chart in order to create the downstream package, using the `render-helm-chart` function. But that function cannot be part of the `kpt` pipeline as it is not hermetic and/or requires volume mounts:

https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/functions/go/render-helm-chart

Reply:
I think of the pipeline included in kpt packages as maintaining certain invariants of the package as it goes through iterations. So it could make sure that all resources will always belong in a particular namespace since it will be automatically run by porch after every update to the package.

I think of a pipeline that is part of the PackageVariant controller as more of one-off updates to the package. So we are kinda automating the manual changes a user might perform on a newly created packagerevision, so basically imperative function invocations. Therefore I'm not sure if I think it makes sense to add the functions to the package pipeline.

Reply:
We can probably make it an option, like `--save`, and leave it up to the user.

Reply:
Updated to reflect this.